### PR TITLE
feat(starry-kernel): expose per-CPU cache, topology and NUMA node sysfs

### PR DESCRIPTION
**Pull request by Lfan-ke**
## 问题

hwloc、pocl、numactl/libnuma 从 sysfs 读取 CPU 缓存层级、CPU 拓扑与 NUMA 节点内存，StarryOS 缺少这些节点。本 PR 对照 Linux `drivers/base/{cacheinfo,topology,node}.c` 补齐。

## 改动

- 每个逻辑 CPU 的 `cache/indexN`：在该 CPU 上读取架构缓存寄存器（x86_64 `CPUID` leaf 4，缺 leaf 4 时按 `intel_cacheinfo_0x2` 回退 leaf 2；aarch64 `CLIDR_EL1`/`CCSIDR_EL1`；loongarch64 `CPUCFG`）。riscv64 没有对应寄存器，Linux 走设备树，这里不暴露 `cache/`。启动时通过同步跨 CPU 调用让每个 CPU 在自己上面读取缓存寄存器，结果写进调用方持有的定长缓冲，调用返回后再在任务上下文里发布到该 CPU 的槽；调用失败的 CPU 槽为空，`cpuN/cache` 不出现，不会用其他 CPU 的寄存器补上。aarch64 在 `CSSELR_EL1` 选择与 `CCSIDR_EL1` 读取期间关抢占，保证两者在同一个核上。
- `shared_cpu_map`/`shared_cpu_list`：私有缓存只列本 CPU；能证明覆盖全部在线 CPU 的共享缓存列出全部在线 CPU；x86 上无法构造完整共享集合的缓存省略这两个属性，既不报成私有，也不伪造成全系统共享。
- `topology/`：`core_id`、`physical_package_id`、`core_cpus`、`core_siblings`、`thread_siblings`、`package_cpus` 及对应 `_list`。`cluster_*` 只在 x86_64/aarch64/riscv64 出现，`die_*` 只在 x86_64 出现，与 Linux 的 `TOPOLOGY_CLUSTER_SYSFS`、`TOPOLOGY_DIE_SYSFS` 一致；目录列举和直接查找受同一门控。
- CPU 位掩码按 `cpu_num()` 生成 32 位分组（对应 `cpumap_print_to_pagebuf`），超过 64 个 CPU 时不截断。
- `node/node0`：`meminfo` 取与 `/proc/meminfo` 同源的分配器统计，`MemUsed == MemTotal - MemFree`；`distance` 为 10；`cpulist` 与 `cpu/{online,possible,present}` 恰好一个结尾换行。

## 本轮更新

- **axtest 编译失败**：CI 四个架构的 Starry 套件失败，原因是 `sysfs.rs` 与 `sysfs/cache.rs` 中两个只在宿主运行的测试模块使用 `#[cfg(test)]`，内核 axtest 构建里 `use super::*` 成了未使用的导入，在 `-D warnings` 下编译失败；x86_64 日志中其余 507 项用例都已通过。两个模块改为仓库惯例 `#[cfg(all(test, not(axtest)))]`。
- **跨 CPU 采样回调不再分配**：`run_on_cpu_sync` 最终调用 `axipi` 的 `call_on_cpu`，其契约要求回调在硬中断上下文中执行，不得睡眠、分配、释放持有的状态，也不得获取调用方可能持有的锁。上一轮以分配器的锁都是 `lock_irqsave` 为由在回调里构造 `Vec` 并替换槽内旧值，但契约禁止的是分配与释放本身，与锁是否中断安全无关，这一说法不成立。现在三个架构的读取函数都写入容量为 16 的 `heapless::Vec`（x86 leaf 4 在此截断，leaf 0x2 最多 4 项，arm64 最多 7 级、loongarch64 最多 3 级，每级至多数据与指令两项），回调只把结果写进调用方栈上的缓冲，调用返回后由 `init_cpu_cache` 在任务上下文中转为 `Vec` 发布，回调里既不分配也不取锁。这一性质由类型保证；内核没有在硬中断中检测分配的机制，运行时证据是 SMP 下的 axtest 与拓扑用例。
- **node meminfo 漏算任务栈**：`/proc/meminfo` 计入 `UsageKind::TaskStack`，node 的 `meminfo` 没有，二者对同一块内存给出不同的 `MemFree`。`/proc/meminfo`、`/proc/vmstat` 与 node `meminfo` 的求和改为共用 `allocator_used_bytes`。新增系统测试 `test-sysfs-node-meminfo`：先起 64 个阻塞在管道上的线程，每个占用一块 256 KiB 的内核任务栈，再背靠背读取 `/proc/meminfo` 与 node0 的 `MemFree`，要求 node0 落在两次 `/proc` 读数之间，容差 1 MiB。宿主 Linux 6.6 上两者相等，9 项全过。
- **保持现状的两处**：Linux `drivers/base/node.c` 中 node 的 `possible`、`online`、`has_normal_memory`、`has_cpu` 都经 `%*pbl` 输出节点掩码列表，单节点机器上均为 `0`，宿主 Linux 实测相同，因此四个属性保持同一内容。`CPUID(2)` 按 Linux 7.1 `arch/x86/include/asm/cpuid/api.h` 中 `cpuid_leaf_0x2()` 的处理：迭代次数不等于 1 即视为无效，不做多次迭代。
- 此前一轮已将 `sysfs-info` 的构建与 QEMU 配置从 dev 已删除的 `ax-driver/virtio-blk` 迁到 `ax-driver/nvme`，系统测试改用 `../common` 的测试框架。

## 验证

基线 dev `b101144f`：

| 命令 | 结果 |
|:--:|:--:|
| `cargo xtask starry test qemu --arch x86_64 -c qemu/system/test-sysfs-node-meminfo` | 仅加测试时 3 项失败，node0 的 `MemFree` 比 `/proc/meminfo` 高约 25.8 MB；共用统计后 9 项全过 |
| `cargo xtask starry test qemu --arch x86_64 -c qemu/system/test-sysfs-cpu-topology` | 60 项全过 |
| `cargo xtask ktest qemu -p starry-kernel --arch x86_64` | 192 项全过 |
| `cargo xtask ktest qemu -p starry-kernel --arch aarch64` | 201 项全过 |
| `cargo xtask clippy --package starry-kernel` | 92 项检查全部通过 |
| `cargo xtask test` | 全部通过 |

### CODE REVIEW

**Comment by Lfan-ke** [comment]
Closing to supplement test coverage before merge. The emitted sysfs topology/cache/NUMA-node content is not yet exercised by a companion on-target test. Will reopen with a 4-arch carpet that byte-compares every path against computed golden values (node MemTotal, comma-grouped cpumask, cache attributes) plus format_cpu_mask boundary cases and error-path (ENOENT) negative controls.

File: os/StarryOS/kernel/src/pseudofs/sysfs.rs [hunk 1]
```diff
@@ -767,6 +884,151 @@ impl SimpleDirOps for SystemCpuEntryDir {
...
+                "number_of_sets",
+                "physical_line_partition",
+            ]
+            .into_iter()
+            .map(Cow::Borrowed),
+        )
+    }
+
+    fn lookup_child(&self, name: &str) -> VfsResult<NodeOpsMux> {
+        // (level, type, size KiB, line size, sets, partition) - plausible per-level values.
```
Comment (ZR233) [inline os/StarryOS/kernel/src/pseudofs/sysfs.rs]: 【阻塞｜硬件 ABI 数据错误】这里明确写的是“plausible”值，而不是当前 CPU 的真实 cache 信息；下方还把所有 index（包括通常私有的 L1/L2）的 `shared_cpu_map` 都设为全部在线 CPU。hwloc 会把这些文件当作权威拓扑，得到错误的 cache 共享关系、容量和 sets。请从固件/架构可验证信息派生每 CPU/cache 的 level/type/size/line/sets/shared mask，并让 `CpuCacheIndexDir` 携带 CPU 身份；无法可靠获取的属性应省略，不能猜默认值。

File: os/StarryOS/kernel/src/pseudofs/sysfs.rs [hunk 2]
```diff
@@ -697,11 +697,124 @@ impl SimpleDirOps for SystemDir {
...
+            _ => return Err(VfsError::NotFound),
+        })
+    }
+}
+
+/// `Node 0 MemTotal: …` block that hwloc parses to learn per-node memory.
+fn render_node_meminfo() -> String {
+    let total_kb = ax_runtime::hal::mem::total_ram_size() / 1024;
+    format!(
+        "Node 0 MemTotal:       {total_kb:>10} kB\nNode 0 MemFree:        {total_kb:>10} kB\nNode \
```
Comment (ZR233) [inline os/StarryOS/kernel/src/pseudofs/sysfs.rs]: 【阻塞｜动态内存统计错误】这里无论运行时已经分配多少内存，都把 `MemFree` 写成 `MemTotal`、`MemUsed` 写成 0。用户态会把 sysfs node meminfo 当作真实状态，这会产生互相矛盾的容量/使用量。请接入可验证的 allocator/内存统计；如果当前只能可靠提供 MemTotal，就只暴露能保证正确的字段，并用 guest 测试锁定格式与数值关系。

**Review by ZR233 — CHANGES_REQUESTED** [review]
本 PR 为 `/sys/devices/system/{cpu,node}` 增加 topology、cache 与 node0 meminfo 节点，目标是让 hwloc/pocl 完成 Linux sysfs discovery。当前 head 的格式、clippy、四架构 Starry QEMU 与相关 board CI 均通过；互斥 host/container skip 属预期矩阵行为，因此未重复这些宽泛 CI 等价命令。`git diff --check origin/dev...HEAD` 通过。

阻塞问题：实现把 cache 层级、容量、line size、sets、package/core/die 关系硬编码成“plausible”值，并将每个 cache（包括 per-core L1/L2）都报告为全 CPU 共享。这不是 Linux sysfs 的真实硬件拓扑语义，会让正是本 PR 要支持的 hwloc/pocl 基于错误数据做线程/内存调度。项目平台规则不允许猜测硬件元数据；应从固件/架构 cache/topology 信息派生，无法可靠得到的属性应不暴露，而不是伪造。node0 meminfo 同时把 MemFree 永远报告为 MemTotal、MemUsed 永远为 0，也会向用户态提供错误的动态内存状态。

测试覆盖同样阻塞：本 PR 新增用户可见 ABI，但没有任何测试文件，PR body 和 commit message 也没有可复现的非板卡验证命令。现有 aggregate CI 只能证明系统未普遍回归，不能证明这些新 sysfs 文件的内容、格式和 hwloc/pocl 后置条件。请新增正确层级的 Starry test-suit 回归，至少读取各 CPU 的 topology/cache 与 node0 文件，校验 mask/list 格式和拓扑一致性，并运行实际 hwloc discovery 或等价最小解析器证明内存非零；在 PR 描述中给出精确 xtask 命令和 guest-visible 结果。

重复/重叠：已检查 `origin/dev` 和开放 PR，base 没有等价的 SystemNodeDir/CpuCacheDir/topology 属性，开放 PR 中没有其他 hwloc/topology/cache/NUMA 实现，因此不是重复或被替代的工作。

历史两个 rustfmt 评论在当前 head 已由格式 CI 覆盖，我已解析并回读确认 `isResolved=true`。结论：方向合理，但当前伪造硬件/内存元数据且缺少功能回归，按 Linux ABI 与项目测试门槛请求修改。

**Comment by Lfan-ke** [comment]
Reworked to emit real values instead of the fabricated ones the review flagged, and added a doc-grounded carpet.

- **Cache is now read from the hardware**, not a plausible-looking table: x86_64 CPUID leaf 4 with a leaf 0x2 legacy-descriptor fallback (matching `arch/x86/kernel/cpu/cacheinfo.c` + `cpuid_0x2_table.c`), aarch64 CLIDR_EL1/CCSIDR_EL1, loongarch64 CPUCFG. riscv64 has no cache-geometry registers, so `cache/` is omitted exactly as Linux does when cacheinfo is unavailable, rather than inventing entries. `shared_cpu_map`/`shared_cpu_list` now follow `cache_leaves_are_shared` (L1 private, L2+ system-wide), and per-attribute visibility mirrors `cache_default_attrs_is_visible` (unknown geometry fields are absent, not zero).
- **node meminfo now reuses the live page-allocator accounting** shared with `/proc/meminfo`, so `MemUsed == MemTotal - MemFree` holds and the two views cannot disagree (matches `drivers/base/node.c`). Fixed an off-by-one where the three fields were each truncated independently.
- Added the missing topology attrs Linux always exposes (`core_cpus_list`/`package_cpus_list`/`thread_siblings`[`_list`]).
- `apps/starry/sysfs-info` reads and asserts the real values on all four arches: x86_64 52/52, aarch64 45/45, riscv64 19/19 (cache correctly absent), loongarch64 52/52, all TEST PASSED single-core. x86_64 uses `-cpu Haswell -accel tcg` because nested KVM here masks the cache CPUID leaves; the Haswell TCG model exposes them. Observed x86 cache: L1D/L1I 32K/8-way, L2 4M/16-way, L3 16M/16-way, geometry identity `size == sets*line*ways` verified on aa/la/x86.

fmt clean, clippy 20/20.

File: os/StarryOS/kernel/src/pseudofs/sysfs.rs [hunk 3]
```diff
@@ -767,6 +913,679 @@ impl SimpleDirOps for SystemCpuEntryDir {
...
+
+/// `/sys/devices/system/cpu/cpu<N>/cache/` — the real cache leaves enumerated for this CPU.
+struct CpuCacheDir {
+    fs: Arc<SimpleFs>,
+    cpu: usize,
+}
+
+impl SimpleDirOps for CpuCacheDir {
+    fn child_names<'a>(&'a self) -> Box<dyn Iterator<Item = Cow<'a, str>> + 'a> {
+        let n = read_cache_leaves().len();
```
Comment (ZR233) [inline os/StarryOS/kernel/src/pseudofs/sysfs.rs]: 【阻塞｜CPU 身份错误】`CpuCacheDir` 保存了 `self.cpu`，这里却调用无参数的 `read_cache_leaves()`，实际读取的是执行这次 sysfs 操作的当前 CPU。于是访问 `/cpu0/cache` 可能得到调度该读取的任意核信息，big.LITTLE 上各个 `cpu<N>` 会重复或随调度变化。请在各 CPU bring-up 时固定发现结果，或保证在目标 CPU 上执行并按 CPU 缓存；增加两个异构 CPU leaf 不同的回归。

File: os/StarryOS/kernel/src/pseudofs/sysfs.rs [hunk 4]
```diff
@@ -767,6 +913,679 @@ impl SimpleDirOps for SystemCpuEntryDir {
...
+    }
+    let ccidx = ((ccidx_field >> 20) & 0xf) != 0;
+
+    // Read CCSIDR for leaf (level, instruction-side?), restoring CSSELR afterwards.
+    let read_ccsidr = |level: u32, instr: bool| -> u64 {
+        let sel = ((level as u64 - 1) << 1) | (instr as u64);
+        let saved: u64;
+        let val: u64;
+        unsafe {
+            asm!("mrs {}, csselr_el1", out(reg) saved, options(nomem, nostack, preserves_flags));
```
Comment (ZR233) [inline os/StarryOS/kernel/src/pseudofs/sysfs.rs:1561]: 【阻塞｜迁移破坏寄存器序列】`CSSELR_EL1` 的 save/select/`CCSIDR_EL1` read/restore 必须发生在同一 PE；当前序列没有禁抢占/绑核保证。若在 `msr csselr_el1` 后迁移，会从另一核按未选择的 leaf 读取，并把原核保存值写到新核。请把完整序列置于明确的不可迁移临界区，或只在 per-CPU 初始化阶段读取并缓存，同时记录这一安全不变量。

File: os/StarryOS/kernel/src/pseudofs/sysfs.rs [hunk 5]
```diff
@@ -767,6 +913,679 @@ impl SimpleDirOps for SystemCpuEntryDir {
...
+/// higher levels are shared system-wide.
+struct CpuCacheIndexDir {
+    fs: Arc<SimpleFs>,
+    cpu: usize,
+    leaf: CacheLeaf,
+}
+
+impl CpuCacheIndexDir {
+    /// The set of CPUs sharing this cache: private (this CPU) for L1, system-wide otherwise.
+    fn shared_mask(&self) -> String {
```
Comment (ZR233) [inline os/StarryOS/kernel/src/pseudofs/sysfs.rs:1756]: 【阻塞｜共享拓扑被虚构】“L1 私有、L2+ 全系统共享”不成立：常见多 cluster/多 socket 系统的 L2 只在 cluster 内共享，L3 也未必覆盖所有 online CPU。这里会向用户态暴露错误的 `shared_cpu_map/list`。Linux 是按 firmware/arch cache ID 与拓扑构造 shared map；拿不到可靠拓扑时应省略属性或只报告可证明的集合，不能把所有 CPU 填进去。

**Review by ZR233 — CHANGES_REQUESTED** [review]
上轮伪造 meminfo/静态容量的问题已改为读取 live allocator，current-head 29 个 CI check 通过；但新增 cache sysfs 仍有三处硬件语义阻塞：cpu<N> 没有读取 N 对应 CPU、AArch64 的 CSSELR/CCSIDR 序列允许迁移后跨 CPU 读写选择寄存器、共享拓扑则被简化为“L1 私有、其余全系统共享”。QEMU 单 CPU 覆盖无法发现这些问题；请按 Linux cacheinfo 的 per-CPU 初始化/固化模型保存每个 CPU 的 leaf 与真实 shared mask，并补异构 CPU/迁移场景的回归覆盖。

**Review by ZR233 — CHANGES_REQUESTED** [review]
本轮复审确认，当前 head 已修复上次指出的多项问题：cache 几何不再使用虚构默认值，node meminfo 接入了真实内存统计，AArch64 的 CSSELR/CCSIDR 读取序列也由 `NoPreemptIrqSave` 保护；对应过时线程可以关闭。格式线程此前已经解决。

仍有两个既有阻塞没有闭环。`CpuCacheDir` 虽保存 `self.cpu`，`child_names()` 与 `lookup_child()` 仍调用无参数的 `read_cache_leaves()`，读取执行 sysfs 请求的当前 PE，而不是路径中的 `cpu<N>`；在异构 CPU 上会让 `/cpu1/cache` 暴露 CPU0 的 leaves，若两次调用间迁移，目录项与几何还可能不一致。`shared_cpu_map/list` 则从“L2+ 全系统共享”的猜测改成“只包含 owner CPU”，但对真实共享 cache 仍是错误 ABI；未知拓扑不能作为准确的私有关系暴露。请使用 per-CPU cacheinfo/目标 CPU 执行结果和可靠的 firmware/arch cache ID 拓扑；若属性无法可靠确定，应遵循可见性规则省略，而不是返回猜测。当前两条未过时的 ZR233 线程分别位于 `sysfs.rs:1431` 与 `sysfs.rs:1473`，继续保留为阻塞。

验证：`git diff --check origin/dev...HEAD` 通过；`cargo xtask starry app qemu -t sysfs-info --arch aarch64` 通过 45/45 并打印 `SYSFS_CARPET TEST PASSED`，但明确以 SMP1 启动，不能覆盖上述 CPU 身份/共享拓扑缺陷；SMP4 的 `cargo xtask starry test qemu --arch aarch64 -c qemu/test-proc-mem-monitor` 定向重跑通过并得到 `STARRY_GROUPED_TESTS_PASSED`。current-head Actions 中 AArch64 Starry 的 child RSS/COW 失败与既有 #1418 同类，本地定向重跑通过，暂无证据归因本 PR；其余因 fail-fast 取消的架构也不能算覆盖。新增 sysfs-info app 不在本 PR CI 中，且仍缺少能在旧实现稳定失败的双 CPU 回归。请把回归放入现有 SMP4 `test-suit/starryos/qemu/system`，至少证明 `cpu<N>` 与读取任务所在 CPU 解耦，并覆盖一个已知共享的高层 cache。

未发现功能重复 PR；#1577、#1601–#1603 同时修改 `pseudofs/sysfs.rs`，属于高冲突但不替代本 PR 的并行改动，合并顺序需注意。

**Review by ZR233 — CHANGES_REQUESTED** [review]
本轮复审基于 current head `c4dc89c1018aff547f6ca179ae996441ff63b4be`。结论仍为 **Request changes**。

本次更新解决了“读取 sysfs 的执行 CPU 决定所有条目”的旧实现问题：缓存信息现在先写入 `CPU_CACHE_TABLE`，读取路径不再现场读寄存器。但表的初始化仍只在 CPU0 上执行，然后把 CPU0 的 `CacheLeaf` 完整复制到每个其他 CPU slot：

```rust
let seed = table.first().cloned().unwrap_or_default();
for slot in table.iter_mut().skip(1) {
    if slot.is_empty() {
        *slot = seed.clone();
    }
}
```

因此 PR 标题和 sysfs ABI 所宣称的 **per-CPU topology/cache** 在异构 CPU、big.LITTLE、多 socket 或不同 cache-sharing domain 上仍会返回伪造的同质数据。代码注释也明确说明 secondary CPU bring-up hook 尚未接入；这不是实现细节，而是本功能正确性的前置条件。请在每个 CPU 的本地 bring-up 上采样并发布自己的 cache leaves，或缩小本 PR 的承诺与导出范围，不能用 CPU0 副本填充未知值。

第二个阻塞项是 shared_cpu_map/shared_cpu_list。当前 `SharingScope` 对 L1 以外的 cache（以及 x86 上 `num_threads_sharing > 1`）直接标成所有 online CPUs；但 Linux x86 会结合 sharing count 与 APIC-ID 分组构造同一 cache domain，而不是把“多于一个线程共享”解释成“全系统共享”。参考 current local Linux source `8cd9520d35a6c38db6567e97dd93b1f11f185dc6`：

- generic fallback：[drivers/base/cacheinfo.c](https://github.com/torvalds/linux/blob/8cd9520d35a6c38db6567e97dd93b1f11f185dc6/drivers/base/cacheinfo.c#L40-L57)
- x86 override：[arch/x86/kernel/cpu/cacheinfo.c](https://github.com/torvalds/linux/blob/8cd9520d35a6c38db6567e97dd93b1f11f185dc6/arch/x86/kernel/cpu/cacheinfo.c#L547-L580)

请按实际 sharing domain 构造 mask；在无法确定时，不应把所有 CPU 作为精确答案导出。

### 回归与本地验证

- ✅ `cargo fmt --all --check`
- ✅ runner 能发现 `qemu/system/test-sysfs-cpu-topology`，该 C case 完成配置、编译和 rootfs 安装
- ⚠️ guest runtime 在启动前被审阅环境的 QEMU snapshot `/var/tmp` 只读限制挡住，因此没有把“已编译”当作运行通过
- ⚠️ `cargo xtask clippy --package starry-kernel` 已完成 base、axtest、dev-log、dynamic_debug 等前置组合且未出现代码诊断，但完整 22 组合在会话时限内未完成
- ✅ GitHub 当前显示 mergeable
- ✅ 未发现完全重复 PR；#1577/#1601/#1602/#1603 是同一区域的高冲突补充改动，需要后续协调，但不是本 PR 的替代实现
- ⚠️ 未获得 current-head 的完整 GitHub check-run 列表，本结论不以空状态面替代 CI 证据

新增 SMP4 用例使用同质 QEMU Haswell，并断言被复制后的各 CPU 几何信息相同；它无法让两个 CPU 的本地寄存器结果不同，也无法在旧的“CPU0 复制给所有 slot”实现上稳定变红。请至少增加一个最低层确定性测试，给不同 CPU slot 注入不同 cache geometry/sharing domain，证明旧实现红、修复后绿。

本 PR 修改的是只读 sysfs pseudo-files，没有改变 syscall 参数、errno 或状态转换语义，因此 syscall guideline 在本轮不适用；问题位于 sysfs 导出数据本身的真实性与 Linux topology ABI 兼容性。

**Review by ZR233 — APPROVED** [review]
本轮按 current head `0152bc59680b381a14ffb6fa7a567b48a0e79644` 增量复核，先前阻塞项均已被后续提交修复：

- per-CPU cache leaf 已在目标 CPU 上通过 `run_on_cpu` 采样，不再复制 cpu0；
- 大 CPU 编号的 cpumask 已改为可扩展格式；
- aarch64/loongarch64 的 system 回归现在强制要求 cache tree，不能再以 NOTE 跳过；
- x86 CPUID(2) 的计数处理与当前 Linux `cpuid_leaf_0x2()` 一致；无 firmware cache topology 时的 L2+ sharing fallback 也与 Linux `cache_leaves_are_shared()` 一致。

验证清单：

- [x] current head 可合并，`git diff --check` 通过；
- [x] x86_64/aarch64/riscv64/loongarch64 四个 current-head Starry system job 全部成功；各 job 都构建、安装并由 system runner 遍历 `test-sysfs-cpu-topology`；
- [x] 维护性、unsafe/并发、ABI 与文档检查未发现新的阻塞；
- [x] 8 个已失效旧线程已关闭。

当前没有阻塞项，同意合入。

**Review by ZR233 — CHANGES_REQUESTED** [review]
补充复核：在上一条 review 提交前，current head 新增了一个有效的 x86 cache ABI 阻塞，因此撤回同意结论并改为 request changes。

当 CPUID.4 的 `num_threads_sharing` 大于 1、但小于全部 online CPU 时，当前 `LocalGroup` 分支把 `shared_cpu_map/list` 固定为 owner-only。该值不是保守且正确的 sysfs 表达，而会把 SMT sibling 或局部 cache group 的真实共享 cache 报成私有；Linux x86 通过 APIC/cache topology 计算完整分组。现有四架构 QEMU system job 使用的 x86 拓扑没有覆盖这个局部分组场景，因此全绿不能证明该路径。

更新后的检查结论：

- [x] per-CPU 采样、可扩展 cpumask、架构强制 cache 回归、CPUID(2) 与 generic arch fallback 等旧问题已修复；
- [x] current head 四架构 system CI 成功；
- [ ] x86 `LocalGroup` 的完整 sharer 集合仍错误，current inline thread 保持未解决；
- [ ] 需要增加 SMT/局部共享（不是 private、也不是 system-wide）的确定性回归。

请保存每 CPU 的 APIC/cache grouping 信息并输出完整共享集合；如果无法可靠构造，不能把已知共享的 cache 暴露为 owner-only。

**Comment by Lfan-ke** [comment]
@ZR233
